# --------------------------------------------------------------------------- #
# A Snakemake workflow for the detection of small nucleotide and structural   #
# variants in bacterial ONT long-read sequencing data.                        #
#                                                                             #
# Author: Thomas F. Wulff                                                     #
# Date: 2024-01-15                                                            #
# License: Free for non-commercial use (see LICENSE).                         #
#          For all 3rd party tools separate licenses may apply.               #
# --------------------------------------------------------------------------- #

# --------------------------------------------------------------------------- #
# Set paths and parameters for workflow                                       #
# --------------------------------------------------------------------------- #
configfile: "config/config.yml"
rootdir = os.getcwd()
indir = os.path.join(rootdir, config["indir"])
outdir = os.path.join(rootdir, config["outdir"])

# --------------------------------------------------------------------------- #
# Retrieve sample information from config.yml                                 #
# --------------------------------------------------------------------------- #
include: "rules/common.smk"
SAMPLEINFO = read_sample_sheet()
SAMPLES = list(SAMPLEINFO["sample"])
GENOMES = list_reference_genomes()
wildcard_constraints: sample = "[^/|.]*"

# --------------------------------------------------------------------------- #
# Include sub-workflows                                                       #
# --------------------------------------------------------------------------- #
include: "rules/mapping.smk"
include: "rules/small_nucleotide_variants.smk"
include: "rules/structural_variants.smk"
include: "rules/variant_filtering.smk"
include: "rules/quality_control.smk"
include: "rules/report.smk"

rule all:
    input:
        generate_results